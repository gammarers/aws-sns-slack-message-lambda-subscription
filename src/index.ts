import { ResourceDefaultNaming, ResourceNaming, ResourceNamingType } from '@gammarers/aws-resource-naming';
import { Duration, Names, RemovalPolicy, Stack } from 'aws-cdk-lib';
import * as iam from 'aws-cdk-lib/aws-iam';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import * as logs from 'aws-cdk-lib/aws-logs';
import * as sns from 'aws-cdk-lib/aws-sns';
import { Construct } from 'constructs';
import { SendSlackMessageFunction } from './funcs/send-slack-message-function';

export interface ResourceCustomNaming {
  readonly type: ResourceNamingType.CUSTOM;
  readonly functionName: string;
  readonly functionRoleName: string;
}

export type ResourceNamingOption = ResourceDefaultNaming | ResourceCustomNaming;

export interface SNSSlackMessageLambdaSubscriptionProps {
  readonly topic: sns.ITopic;
  readonly slackWebhookSecretName: string;
  readonly resourceNamingOption?: ResourceNamingOption;
}

export class SNSSlackMessageLambdaSubscription extends Construct {
  constructor(scope: Construct, id: string, props: SNSSlackMessageLambdaSubscriptionProps) {
    super(scope, id);

    // ðŸ‘‡ Get account & region
    const account = Stack.of(scope).account;
    const region = Stack.of(scope).region;

    // ðŸ‘‡ Create random 8 length string
    const random = ResourceNaming.createRandomString(`${Names.uniqueId(scope)}.${Names.uniqueId(this)}`);
    const autoNaming = {
      functionName: `send-slack-message-${random}-func`,
      functionRoleName: `send-slack-message-lambda-func-exec-${random}-role`,
    };
    const names = ResourceNaming.naming(autoNaming, props.resourceNamingOption as ResourceNaming.ResourceNamingOption);

    // Lambda function
    const sendMessageFunction = new SendSlackMessageFunction(this, 'SendSlackMessageFunction', {
      functionName: names.functionName,
      description: 'Send Slack message function',
      architecture: lambda.Architecture.ARM_64,
      role: new iam.Role(this, 'LambdaExecutionRole', {
        roleName: names.functionRoleName,
        description: '',
        assumedBy: new iam.ServicePrincipal('lambda.amazonaws.com'),
        managedPolicies: [
          iam.ManagedPolicy.fromAwsManagedPolicyName('service-role/AWSLambdaBasicExecutionRole'),
        ],
        inlinePolicies: {
          'get-secret-policy': new iam.PolicyDocument({
            statements: [
              new iam.PolicyStatement({
                effect: iam.Effect.ALLOW,
                actions: [
                  'secretsmanager:GetSecretValue',
                ],
                resources: [
                  `arn:aws:secretsmanager:${region}:${account}:secret:${props.slackWebhookSecretName}*`,
                ],
              }),
            ],
          }),
        },
      }),
      timeout: Duration.seconds(10),
      paramsAndSecrets: lambda.ParamsAndSecretsLayerVersion.fromVersion(lambda.ParamsAndSecretsVersions.V1_0_103, {
        cacheSize: 500,
        logLevel: lambda.ParamsAndSecretsLogLevel.INFO,
      }),
      environment: {
        SLACK_WEBOOK_SECRET_NAME: props.slackWebhookSecretName,
      },
      logFormat: lambda.LogFormat.JSON,
      systemLogLevel: lambda.SystemLogLevel.INFO,
      applicationLogLevel: lambda.ApplicationLogLevel.INFO,
      logGroup: new logs.LogGroup(this, 'SendSlackMessageFunctionLogGroup', {
        logGroupName: names.functionName ? `/aws/lambda/${names.functionName}` : undefined,
        retention: logs.RetentionDays.THREE_MONTHS,
        removalPolicy: RemovalPolicy.RETAIN_ON_UPDATE_OR_DELETE,
      }),
    });
    sendMessageFunction.addPermission('InvokeFunctionFromSNS', {
      principal: new iam.ServicePrincipal('sns.amazonaws.com'),
      action: 'lambda:InvokeFunction',
      sourceArn: props.topic.topicArn,
    });

    new sns.Subscription(this, 'LambdaSubscription', {
      topic: props.topic,
      protocol: sns.SubscriptionProtocol.LAMBDA,
      endpoint: sendMessageFunction.functionArn,
    });
  }
}
