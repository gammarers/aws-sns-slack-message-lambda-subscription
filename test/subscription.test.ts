import { App, Stack } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import * as sns from 'aws-cdk-lib/aws-sns';
import { SNSSlackMessageLambdaSubscription } from '../src';

describe('Subscription Testing', () => {

  const app = new App();
  const stack = new Stack(app, 'TestingStack', {
    env: {
      account: '123456789012',
      region: 'us-east-1',
    },
  });

  const topic = new sns.Topic(stack, 'Topic', {
    topicName: 'example-topic',
  });

  new SNSSlackMessageLambdaSubscription(stack, 'SNSSlackMessageLambdaSubscription', {
    topic,
    slackWebhookSecretName: 'slak-webhook',
  });

  const template = Template.fromStack(stack);

  it('Is SNS Subscription Resource Count', () => {
    template.resourceCountIs('AWS::SNS::Subscription', 1);
  });

  it('Is Lambda Function Resource Count', () => {
    template.resourceCountIs('AWS::Lambda::Function', 1);
  });

  it('Is IAM Role Resource Count', () => {
    template.resourceCountIs('AWS::IAM::Role', 1);
  });

  it('Is Logs LogGroup Resource Count', () => {
    template.resourceCountIs('AWS::Logs::LogGroup', 1);
  });

  it('Should match snapshot', () => {
    expect(template.toJSON()).toMatchSnapshot();
  });
});