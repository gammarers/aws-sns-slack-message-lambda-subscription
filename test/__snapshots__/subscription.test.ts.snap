// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Subscription Testing Should match snapshot 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "SNSSlackMessageLambdaSubscription5EE4F2B8": {
      "Properties": {
        "Endpoint": {
          "Fn::GetAtt": [
            "SNSSlackMessageLambdaSubscriptionSendSlackMessageFunctionA36ABB02",
            "Arn",
          ],
        },
        "Protocol": "lambda",
        "TopicArn": {
          "Ref": "TopicBFC7AF6E",
        },
      },
      "Type": "AWS::SNS::Subscription",
    },
    "SNSSlackMessageLambdaSubscriptionLambdaExecutionRoleC5480D3F": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
        "Policies": [
          {
            "PolicyDocument": {
              "Statement": [
                {
                  "Action": "secretsmanager:GetSecretValue",
                  "Effect": "Allow",
                  "Resource": "arn:aws:secretsmanager:us-east-1:123456789012:secret:slak-webhook*",
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "get-secret-policy",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "SNSSlackMessageLambdaSubscriptionSendSlackMessageFunctionA36ABB02": {
      "DependsOn": [
        "SNSSlackMessageLambdaSubscriptionLambdaExecutionRoleC5480D3F",
      ],
      "Properties": {
        "Architectures": [
          "arm64",
        ],
        "Code": {
          "S3Bucket": "cdk-hnb659fds-assets-123456789012-us-east-1",
          "S3Key": "HASH.zip",
        },
        "Description": "Send Slack message function",
        "Environment": {
          "Variables": {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "PARAMETERS_SECRETS_EXTENSION_CACHE_ENABLED": "true",
            "PARAMETERS_SECRETS_EXTENSION_CACHE_SIZE": "500",
            "PARAMETERS_SECRETS_EXTENSION_HTTP_PORT": "2773",
            "PARAMETERS_SECRETS_EXTENSION_LOG_LEVEL": "info",
            "PARAMETERS_SECRETS_EXTENSION_MAX_CONNECTIONS": "3",
            "SECRETS_MANAGER_TIMEOUT_MILLIS": "0",
            "SECRETS_MANAGER_TTL": "300",
            "SLACK_WEBOOK_SECRET_NAME": "slak-webhook",
            "SSM_PARAMETER_STORE_TIMEOUT_MILLIS": "0",
            "SSM_PARAMETER_STORE_TTL": "300",
          },
        },
        "Handler": "index.handler",
        "Layers": [
          "arn:aws:lambda:us-east-1:177933569100:layer:AWS-Parameters-and-Secrets-Lambda-Extension-Arm64:4",
        ],
        "LoggingConfig": {
          "ApplicationLogLevel": "INFO",
          "LogFormat": "JSON",
          "LogGroup": {
            "Ref": "SNSSlackMessageLambdaSubscriptionSendSlackMessageFunctionLogGroup791AC086",
          },
          "SystemLogLevel": "INFO",
        },
        "Role": {
          "Fn::GetAtt": [
            "SNSSlackMessageLambdaSubscriptionLambdaExecutionRoleC5480D3F",
            "Arn",
          ],
        },
        "Runtime": "nodejs22.x",
        "Timeout": 10,
      },
      "Type": "AWS::Lambda::Function",
    },
    "SNSSlackMessageLambdaSubscriptionSendSlackMessageFunctionInvokeFunctionFromSNS7E7D00CD": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "SNSSlackMessageLambdaSubscriptionSendSlackMessageFunctionA36ABB02",
            "Arn",
          ],
        },
        "Principal": "sns.amazonaws.com",
        "SourceArn": {
          "Ref": "TopicBFC7AF6E",
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "SNSSlackMessageLambdaSubscriptionSendSlackMessageFunctionLogGroup791AC086": {
      "DeletionPolicy": "RetainExceptOnCreate",
      "Properties": {
        "RetentionInDays": 90,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Retain",
    },
    "TopicBFC7AF6E": {
      "Properties": {
        "TopicName": "example-topic",
      },
      "Type": "AWS::SNS::Topic",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
